Tasks/ExtractFilesV1/extractfilestask.ts

```diff
@@ -20,18 +21,33 @@ tl.debug('win: ' + win);
 var xpTarLocation: string;
 var xpUnzipLocation: string;
 // 7zip
-var xpSevenZipLocation: string;
-var winSevenZipLocation: string = path.join(__dirname, '7zip/7z.exe');
+let xpSevenZipLocation: string;
+let defaultWinSevenZipLocation: string = path.join(__dirname, '7zip/7z.exe');
+let winSevenZipLocation: string;
 
 function getSevenZipLocation(): string {
+    if (customPathToSevenZipTool) {
+        tl.debug('Get 7z tool from user defined location');
+        xpSevenZipLocation = customPathToSevenZipTool;
+        winSevenZipLocation = customPathToSevenZipTool;
+    }
+    
     if (win) {
-        return winSevenZipLocation;
-    } else {
-        if (typeof xpSevenZipLocation == "undefined") {
-            xpSevenZipLocation = tl.which('7z', true);
+        if (!winSevenZipLocation) {
+            tl.debug('Try to resolve preinstalled 7z location');
+            winSevenZipLocation = tl.which('7z', false);
         }
-        return xpSevenZipLocation;
+
+        // return default location of the 7z which is bundled with the task in case the user didn't pass a custom path or the agent doesn't contain a preinstalled tool
+        return winSevenZipLocation || defaultWinSevenZipLocation;
     }
+
+    if (!xpSevenZipLocation) {
+        tl.debug('Try to resolve preinstalled 7z location');
+        xpSevenZipLocation = tl.which('7z', true);
```

Comment:
Could you please clarify why the check option is enabled here and disabled for Windows?

Reply:
It was done because we don't want to fail the task on Windows in case 7z is not preinstalled in the system, since in this case we just return the path to the 7z that comes with the task. MacOS and Linux always relies only on 7z from vm-images so this check is necessary.